src/Storages/System/StorageSystemZooKeeperInfo.cpp

```diff
@@ -0,0 +1,392 @@
+    ConfigProcessor::registerEmbeddedConfig("config.xml", "<clickhouse/>");
+    auto clickhouse_config = config_processor.loadConfig();
+
+    String cluster_name = "default";
+
+    Poco::AutoPtr<Poco::XML::NodeList> zookeeper_elements = clickhouse_config.preprocessed_xml->getElementsByTagName("zookeeper");
+    for (size_t i = 0; i < zookeeper_elements->length(); ++i)
+    {
+        auto* zookeeper_element = dynamic_cast<Poco::XML::Element*>(zookeeper_elements->item(i));
+        if (zookeeper_element)
+        {
+            cluster_name = zookeeper_element->getAttribute("name");
+        }
+    }
+
+    assert(zookeeper_elements->length() == 1);
+
+    String prefix = "zookeeper.node";
+
+    Poco::AutoPtr<Poco::XML::NodeList> nodes = zookeeper_elements->item(0)->childNodes();
+
+    for (size_t i = 0; i < nodes->length(); ++i)
+    {
+        if (nodes->item(i)->nodeName() != "node")
+            continue;
+
+        if (!clickhouse_config.configuration->has(prefix + ".host") ||
+            !clickhouse_config.configuration->has(prefix + ".port"))
+            continue;
+
+        String host = clickhouse_config.configuration->getString(prefix + ".host");
+        String port = clickhouse_config.configuration->getString(prefix + ".port");
+
+        if (clickhouse_config.configuration->has(prefix + ".secure"))
+            host = "secure://" + host;
+
+
+        res_columns[0]->insert(cluster_name);
+        res_columns[1]->insert(host);
+        res_columns[2]->insert(stoi(port));
+
+        int index = 0;
+
+        auto* node_element = dynamic_cast<Poco::XML::Element*>(nodes->item(i));
+        if (node_element)
+        {
+            std::string attribute_value = node_element->getAttribute("index");
+            if (!attribute_value.empty())
+                index = stoi(attribute_value);
+        }
+
+        res_columns[3]->insert(index);
+
+
+        /// ruok command
+        String ruok_output = sendFourLetterCommand(host, port, "ruok");
+        if (ruok_output == "imok")
+            res_columns[4]->insert(true);
+        else
+            res_columns[4]->insert(false);
+
+
+        /// isro command
+        /// The server will respond with "ro" if in read-only mode or "rw" if not in read-only mode.
+        String isro_output = sendFourLetterCommand(host, port, "isro");
+        if (isro_output == "ro")
+            res_columns[5]->insert(true);
+        else
+            res_columns[5]->insert(false);
+
+        /// mntr command
+        String mntr_output = sendFourLetterCommand(host, port, "mntr");
+        std::map <String, String> mntr_responses_map = getTokens(mntr_output, '\t');
+
+        /// /* 6 */{"version", std::make_shared<DataTypeString>(), "The ZooKeeper version."},
+        res_columns[6]->insert(mntr_responses_map["zk_version"]);
+
+        // /* 7 */ {"avg_latency", std::make_shared<DataTypeUInt64>(), "The average latency."},
+        res_columns[7]->insert(stoi(mntr_responses_map["zk_avg_latency"]));
```

Comment:
We need to validate output from commands. (see comment in `sendFourLetterCommand`). At least check that mntr_responses_map has key `zk_avg_latency` and than try parse to int. If it will fail insert NULL.